[SYCL] Implement sub-group mask extension

vladimirlaz opened the pull request:
The specification is available under https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/SubGroupMask/SubGroupMask.asciidoc

Complementary test changes are available under intel/llvm-test-suite#441, intel/llvm-test-suite#462

dm-vodopyanov reviewed (approved):
SYCL RT changed LGTM

vladimirlaz commented:
@intel/dpcpp-specification-reviewers, could you please have a look?

gmlueck commented on sycl/doc/extensions/GroupMask/GroupMask.asciidoc:
```diff
@@ -286,25 +285,24 @@ struct group_mask {
+  group_mask &operator |=(const group_mask& rhs);
+  group_mask &operator ^=(const group_mask& rhs);
+  group_mask &operator <<=(size_t n);
+  group_mask &operator >>=(size_t n);
```
None of these changes are reflected in the table that describes the member functions.  The sample header and the table should be in sync.

vladimirlaz replied:
fixed

gmlueck commented on sycl/doc/extensions/GroupMask/GroupMask.asciidoc:
```diff
@@ -286,25 +285,24 @@ struct group_mask {
-group_mask operator &(const group_mask& lhs, const group_mask& rhs);
-group_mask operator |(const group_mask& lhs, const group_mask& rhs);
-group_mask operator ^(const group_mask& lhs, const group_mask& rhs);
-
```
Do you know what the rationale was originally for defining these as free functions vs. member functions?  What was the reason for changing it?

Tagging @rolandschulz who is better versed on defining operators as hidden friend functions vs. member functions.

vladimirlaz replied:
I got into trouble for implementing free functions as they cannot be implemented in headers. For the host code, it fires immediately (multiple definitions) for the device code it is a problem with complex cases. Implementing device lib for them looks like overkill.

vladimirlaz replied:
I moved to the free friend functions

gmlueck commented on sycl/doc/extensions/GroupMask/GroupMask.asciidoc:
```diff
@@ -259,6 +257,7 @@ struct group_mask {
   };
 
   static constexpr size_t max_bits = /* implementation-defined */;
+  static constexpr size_t marray_size = max_bits/sizeof(uint32_t)/CHAR_BIT;
```
This is only correct if `max_bits` is evenly divisible by `sizeof(uint32_t)/CHAR_BIT`.  The spec does not guarantee that this is the case.  You could update the math here to round up in the case when it is not evenly divisible, or you could replace the math with a comment explaining the intent.

Also, there should be a description of `marray_size` above in the section "Group Masks".  Perhaps add a new sentence after this one:

> The maximum number of bits that can be stored in a group_mask is exposed as a static member variable, `group_mask::max_bits`.

Or, create a new table defining the two constexpr member variables.

vladimirlaz replied:
marray_size is implementation details and will be hidden from a user.
The code was updated to round up.

gmlueck replied:
If `marray_size` is an implementation detail, it cannot be part of the public API as you have now.  So, if you intend it to be an implementation detail, you should change the public member functions to not reference it, and you can delete the definition in the sample header.  There is no need to document something that is an implementation detail.

gmlueck replied:
I still think this variable needs to be defined in the spec since it is part of the public API.  I'd suggest adding a new table before the one that describes the member functions:

```
|===
|Member Variable|Description
|`static constexpr size_t max_bits`
|The maximum number of bits that can be stored in the `group_mask`. 

|`static constexpr size_t marray_size`
|The number of `uint32_t` elements in an `marray` which can represent all of
 the bits in the `group_mask`.

|===
```

And then delete the existing sentence that describes `max_bits`.

Pennycook replied:
Hm.  I don't think I gave this enough thought when setting the default argument.  Originally all of the masks were 128 bits, and so `marray<uint32_t, 4>` was universal; now we have to worry about rounding, things aren't quite as nice.

There might even be cases where this default argument would be wasteful -- for example, if `max_bits` is 8, then every representation of the mask as an `marray<uint32_t, 1>` is going to waste 24 bits.

Maybe it would be better to make this an `marray<char, 1>` instead, and say that `max_bits` must always be a multiple of `CHAR_BIT`?

gmlueck replied:
That sounds reasonable to me.

gmlueck commented:
> @intel/dpcpp-specification-reviewers, could you please have a look?

I added some comments just about the changes to the spec.  It would be better to wait for a review from @Pennycook, though.  He was involved more closely with this API, and he would be better to review the implementation.  John is on vacation this week, returning on Monday.

rolandschulz commented on sycl/doc/extensions/GroupMask/GroupMask.asciidoc:
```diff
@@ -178,64 +176,57 @@ work-item with the id `max_local_range()-1`.
 |Toggle the value of the bit corresponding to the specified _id_.
 
-|`bool operator==(group_mask rhs) const`
+|`bool operator==(const group_mask &rhs) const`
```
Why change all these to take `const&`?

vladimirlaz replied:
make reference to avoid copying opblect. Make reference const to avoid modifications of rhs object.

rolandschulz commented on sycl/doc/extensions/GroupMask/GroupMask.asciidoc:
```diff
@@ -178,64 +176,57 @@ work-item with the id `max_local_range()-1`.
-|`group_mask operator &(const group_mask& lhs, const group_mask& rhs)`
-|Return a mask representing the result of performing a bitwise AND of `lhs` and
- `rhs`.
+|`group_mask operator &(const group_mask &rhs) const`
```
I agree it was wrong. But I think this isn't the change we want. We should have all binary operators be friend members rather than members only taking rhs. This makes sure that the operator is correctly symmetric (e.g. for implicit conversion).

vladimirlaz replied:
Could you please provide an example of the use case for implicit conversion? We do not have any constructor for group_mask rather than copy constructor (we can create an object of that type using group_ballot() only).

vladimirlaz replied:
moved back to free friend functions

Pennycook commented on sycl/doc/extensions/SubGroupMask/SubGroupMask.asciidoc:
```diff
@@ -21,7 +21,7 @@ IMPORTANT: This specification is a draft.
 NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by permission by Khronos.
 
-This document describes an extension which adds a `group_mask` type.  Such a mask can be used to efficiently represent subsets of work-items in a group for which a given Boolean condition holds.  Group mask functionality is currently limited to groups that are instances of the `sub_group` class.
+This document describes an extension which adds a `sub_group_mask` type.  Such a mask can be used to efficiently represent subsets of work-items in a sub-group for which a given Boolean condition holds.  Sub-group mask functionality is currently limited to groups that are instances of the `sub_group` class.
```
```suggestion
This document describes an extension which adds a `sub_group_mask` type.  Such a mask can be used to efficiently represent subsets of work-items in a sub-group for which a given Boolean condition holds.
```

I think you can remove this sentence now -- the name makes it clear that it's sub-group specific, and we don't have any plans to make `sub_group_mask` work with other groups.

Pennycook commented on sycl/doc/extensions/SubGroupMask/SubGroupMask.asciidoc:
```diff
@@ -74,25 +74,25 @@ the specification.
 
-The `group_ballot` algorithm converts a Boolean condition from each work-item
-in the group into a group mask.  Like other group algorithms, `group_ballot`
+The `sub_group_ballot` algorithm converts a Boolean condition from each work-item
```
When we discussed renaming yesterday I hadn't actually thought about what to call the free-function.

@rolandschulz, @gmlueck: Do you think we should call this `sub_group_ballot`, or is it okay for `group_ballot` to return a `sub_group_mask` when handed a `sub_group`, leaving the possibility open for `group_ballot` to return something else when handed a `group`?

gmlueck replied:
If we keep the `Group` type as a template parameter, then I think the function should be named `group_ballot()`.  Alternatively, if we decide to name the function `sub_group_ballot()`, then we should hard-code the parameter type to `sub_group`, rather than using a template parameter.  Either seem OK to me.

vladimirlaz commented:
@gmlueck, @rolandschulz, @Pennycook, could you please have a look on recent changes.
main changes:
* Change default value for max_bits to 32 (compiler flag will be added later)
* Remove marray<> from internal layout (leave for insert_bits and extract_bits). Use uint32_t to store bits.
* Make value for insert_bits(const &T in) and extract_bits(const &T out)  deducible by putting in/out to function arguments.
* Rename to sub_group_mask.
* Update spec acordingly.

vladimirlaz commented:
> I think there are a few spec updates missing. GitHub won't let me comment on the diff for some reason, so doing so here:
> 
> The predefined macro section still says `SYCL_EXT_ONEAPI_GROUP_MASK`, whereas it should say `SYCL_EXT_ONEAPI_SUB_GROUP_MASK` after the renaming.
> 
> I thought we'd decided that we'd use `marray<char, max_bits/CHAR_BIT>` instead of `marray<uint32_t, ...>` for large masks, with a note that `max_bits` must be a multiple of `CHAR_BIT`. That would allow you to get rid of `marray_size`. If `max_bits` is small enough that you don't need an `marray` at all, the specification should allow for that too.

@Pennycook It looks like I forgot to upload the commits. 
 - SYCL_EXT_ONEAPI_GROUP_MASK has been renamed:
 - I replaced marray<> with uint32_t as we have max_bits set to 32.
 - I have removed marray_size from the sources and will remove from the doc (forgot it)

vladimirlaz commented:
@Pennycook, @gmlueck, @dm-vodopyanov, @rolandschulz  could you plise review/approve the PR?

Pennycook commented:
> @Pennycook It looks like I forgot to upload the commits.
> 
> * SYCL_EXT_ONEAPI_GROUP_MASK has been renamed:

There's still one instance of `SYCL_EXT_ONEAPI_GROUP_MASK` (Line 54 of the extension documentation).  Otherwise LGTM.

vladimirlaz commented:
> > @Pennycook It looks like I forgot to upload the commits.
> > 
> > * SYCL_EXT_ONEAPI_GROUP_MASK has been renamed:
> 
> There's still one instance of `SYCL_EXT_ONEAPI_GROUP_MASK` (Line 54 of the extension documentation). Otherwise LGTM.

@Pennycook, I have fixed that. Could you please approve the PR?

bader commented:
> The specification is available under https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/SubGroupMask/SubGroupMask.asciidoc

@vladimirlaz, please, update the link to the specification.

vladimirlaz commented:
> > The specification is available under https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/SubGroupMask/SubGroupMask.asciidoc
> 
> @vladimirlaz, please, update the link to the specification.

the spec is moved in the scope of the PR. The comment contains the new location.
